File: src/euler_017.jl
unit =
    ["one" "two" "three" "four" "five" "six" "seven" "eight" "nine" "ten" "eleven" "twelve" "thirteen" "fourteen" "fifteen" "sixteen" "seventeen" "eighteen" "nineteen"]
decimal = ["dummy" "twenty" "thirty" "forty" "fifty" "sixty" "seventy" "eighty" "ninety"]

function int_to_word(n)
    if n == 1000
        result = "onethousand"
    elseif n >= 100
        q, r = divrem(n, 100)
        tail = r > 0 ? "and" * int_to_word(r) : ""
        return unit[q] * "hundred" * tail
    elseif n >= 20
        q, r = divrem(n, 10)
        tail = r > 0 ? int_to_word(r) : ""
        return decimal[q] * tail
    elseif n > 0
        return unit[n]
    end
end

function answer(range)
    reduce(*, map(int_to_word, range)) |> length
end

answer(1:1_000) |> print

File: src/euler_017.hs
unit = ["dummy","one","two","three","four","five","six","seven","eight","nine","ten","eleven","twelve","thirteen","fourteen","fifteen","sixteen","seventeen","eighteen","nineteen"]
decimal = ["dummy","dummy","twenty","thirty","forty","fifty","sixty","seventy","eighty","ninety"]

intToWord :: Int -> String
intToWord 1000 = "onethousand"
intToWord n
  | n >= 100 = unit !! q ++ "hundred" ++ tail
  | n >=  20 = decimal !! q' ++ tail'
  | n >=   0 = unit !! n
  where
    (q , r ) = divMod n 100
    (q', r') = divMod n  10
    tail  = if r  > 0 then "and" ++ intToWord r else ""
    tail' = if r' > 0 then intToWord r' else ""

awnser = sum . map (length . intToWord)
main = putStrLn $ show $ awnser [1..1000]
